File: 00xx/006x/0066-PlusOne/Solution.go
/*  Solution: Single Loop
    Time Complexity:    O(n)
    Space Complexity:   O(n)
 */

func plusOne(digits []int) []int {
  carry := 1
  index := len(digits) - 1
  for carry > 0 || index >= 0 {
    if digits[index] == 9 {
      carry = 0
      digits[index] = 0
    } else {
      digits[index]++
      return digits
    }
    index--
  }
  return append([]int{1}, digits...)
}

File: 00xx/006x/0066-PlusOne/Solution.py
# Solution: Right-to-left
# Time Complexity:  O(n)
# Space Complexity: O(1)

class Solution:
  def plusOne(self, digits: List[int]) -> List[int]:
    carry = 0
    for i in range(len(digits)-1, -1, -1):
      if digits[i] == 9:
        digits[i] = 0
      else:
        digits[i] += 1
        return digits
    return [1] + digits
